File: chocofarm/az/transport.py
#!/usr/bin/env python3
"""
chocofarm/az/transport.py — RedisTransport: the SOLE owner of the AZ parallel-loop redis raw-bytes
wire protocol (audit item K, the Transport ⊥ Pool ⊥ Task split out of `parallel.py`).

This module owns EVERYTHING about how weights and results travel over redis and nothing about the
process pool (that is `worker_pool.py`) or what one worker computes (that is `worker.py`). Concretely
it owns: the bounded-timeout connection construction + fail-loud ping (ADR-0002 / deadlock fix H2),
the ONE place the `az:w:<run>:<version>:m|:b` weight keys and the `az:res:<token>:<idx>:X|PI|M|Y`
result keys are spelled (`weight_keys()` / `result_keys()` — no f-strings scattered across
publish/ensure_net/gen_task/collect anymore), the weight publish/read, the result blob write, and the
result blob read+delete. The TTLs live here too (weights 3600s; results `CHOCO_RESULT_TTL`).

Weight (un)packing STAYS delegated to `WeightContainer` (audit item J): this module calls
`pack_net`/`unpack_net` and never re-encodes the layout. The key STRINGS and the on-the-wire bytes are
byte-identical to the pre-split encoder — the protocol is unchanged, only its ownership is.

Connection facts come from `chocofarm/config.py` (`redis_params()` / `redis_socket_timeout()` /
`redis_connect_timeout()`), defaulting to 127.0.0.1:6379 db 0 — DO NOT change which instance/db this
targets (the module is shared with the registry's `redis_params()`).

Public Domain (The Unlicense).
"""
from __future__ import annotations

import os

import numpy as np


# ---- net (de)serialization as raw bytes — DELEGATED to WeightContainer (audit item J) ----
# The raw-bytes (de)serialization is the WeightContainer's (audit item J): the net's weight LAYOUT —
# the param-key order, the manifest's name/shape/dtype/offset/len entries, the scalar meta, and the
# `tobytes()`/`np.frombuffer` round-trip — has ONE owner there, the same owner the npz save/load and
# the L2-mask route through. The transport NO LONGER re-enumerates `net._params()` into its own
# manifest (that was the split-brained second encoder R11 deferred); it constructs the net from the
# manifest's meta and delegates the (un)packing. The blob/manifest bytes are byte-identical to the
# pre-J encoder (the param order is the container's canonical order, which is the historical order).


def pack_net(net):
    """Pack a ValueMLP into (manifest_json: str, blob: bytes) — DELEGATED to
    `WeightContainer.pack` (the one owner of the weight layout, audit item J). Raw `tobytes()` of each
    weight concatenated, a JSON manifest of (name, shape, dtype, offset, byte-length) + the scalar
    meta. No pickle: the blob is contiguous float64 weight bytes; optional params (residual block) ride
    along automatically because the container reports them iff the net built the block."""
    from chocofarm.az.weights import WeightContainer
    return WeightContainer.pack(net)


def unpack_net(manifest_json, blob):
    """Reconstruct a ValueMLP from `pack_net`'s (manifest, blob). The container reads the manifest's
    construction meta (so older manifests without `residual` → block OFF); this builds the net via the
    `ValueMLP` constructor (so the Wr*/br* layout entries have a slot to bind to), then the container
    binds the arrays (`np.frombuffer` views, copied so the net owns writable arrays). No pickle."""
    from chocofarm.az.mlp import ValueMLP
    from chocofarm.az.weights import WeightContainer
    m = WeightContainer.unpack_meta(manifest_json)
    net = ValueMLP(m["in_dim"], hidden=m["H"],
                   n_actions=m["n_actions"], y_mean=m["y_mean"], y_std=m["y_std"],
                   residual=bool(m.get("residual", False)))
    return WeightContainer.unpack_into(net, manifest_json, blob)


# ---- the wire protocol's KEY NAMESPACE (the ONE place the key strings are spelled) ----
# These build the byte-identical keys the pre-split code had inlined as f-strings in publish_weights /
# _ensure_net / _gen_task / _collect_results. Putting them here makes RedisTransport the sole owner of
# the on-the-wire protocol: a key change is a one-site change, and the worker side (worker.py) builds
# its read/write keys through the SAME helpers, so the parent and child can never disagree.

def weight_keys(run, version):
    """The two weight keys for (run, version): (manifest_key, blob_key) =
    `az:w:<run>:<version>:m`, `az:w:<run>:<version>:b`. Byte-identical to the pre-split f-strings."""
    base = f"az:w:{run}:{version}"
    return base + ":m", base + ":b"


def result_keys(res_token, idx):
    """The four result-blob keys for (res_token, idx): (X, PI, M, Y) =
    `az:res:<token>:<idx>:X|PI|M|Y`. Byte-identical to the pre-split f-strings."""
    base = f"az:res:{res_token}:{idx}"
    return base + ":X", base + ":PI", base + ":M", base + ":Y"


# ---- redis connection (raw-bytes transport; no pickle) ----
def _redis_params():
    """Shared connection facts from chocofarm/config.py (the registry uses the same), so transport
    and registry address one redis instance by default."""
    from chocofarm import config
    return config.redis_params()


def connect():
    import redis  # local import so a serial (workers=0) run needs no redis at all
    # Bound EVERY socket op (ADR-0002 / deadlock fix H2). The default `socket_timeout=None`
    # makes every r.get / pipe.execute block FOREVER if the TCP socket stalls — a stalled
    # worker read is then indistinguishable, from the parent's imap fan-out, from a wedged
    # worker, and the loop sits at futex_do_wait at ~1% CPU with no way out. A bounded timeout
    # turns a stall into a loud redis.TimeoutError (retryable / restart-recoverable; checkpoints
    # are per-iteration) instead of a silent permanent hang. Loopback redis under no memory
    # pressure never trips 60s, so this is a safety net, not a happy-path behavior change.
    from chocofarm import config
    r = redis.Redis(
        socket_timeout=config.redis_socket_timeout(),
        socket_connect_timeout=config.redis_connect_timeout(),
        **_redis_params(),
    )
    r.ping()   # fail loud now if redis is unreachable (ADR-0002), not mid-iteration
    return r


# TTL for result blobs: a TTL so an ABORTED iteration self-cleans. The happy path deletes them in
# `read_and_delete_results` the same iteration; but if the fan-out is aborted (the loud drain timeout)
# the parent never reaches the delete, and a bare SET leaves the blob with no expiry forever (the
# post-mortem found ~980 such leaked az:res:* keys, TTL=-1). A 1h TTL bounds that leak without
# affecting the happy path (read+deleted within seconds). `ex=` sets the expiry in the same SET
# round-trip (no extra command). Env-overridable.
def _result_ttl():
    return int(os.environ.get("CHOCO_RESULT_TTL", "3600"))


# Weight keys carry a 1h expiry so a long run doesn't leak old versions.
_WEIGHT_TTL_S = 3600


class RedisTransport:
    """The SOLE owner of the AZ parallel-loop redis raw-bytes protocol. Construct on the parent (with
    a `connect()`'d client) for weight publish + result read; the worker side calls the module-level
    read/write functions with its OWN connection (kept in `_W`, item L) — but the key strings, the
    TTLs, and the (un)packing all route through this module either way, so there is exactly one wire
    protocol."""

    def __init__(self, conn):
        self.r = conn

    # ---- weights: parent publishes, worker reads ----
    def publish_weights(self, net, version, run):
        """Pack the net to raw bytes and publish to redis `az:w:<run>:<version>` (no pickle, no disk).
        Workers `read_weights` it when the version changes. Weight keys carry a 1h expiry."""
        manifest, blob = pack_net(net)
        mk, bk = weight_keys(run, version)
        pipe = self.r.pipeline(transaction=False)
        pipe.set(mk, manifest)
        pipe.set(bk, blob)
        pipe.expire(mk, _WEIGHT_TTL_S)
        pipe.expire(bk, _WEIGHT_TTL_S)
        pipe.execute()

    # ---- results: worker writes, parent reads+deletes ----
    def read_and_delete_results(self, res_token, metas):
        """Read the raw-byte result blobs the workers wrote for `metas` (a list of (idx, n, feat_dim,
        n_slots)) back into one flat list of (feat, pi, mask, g) records, then DELETE the keys (raw
        bytes can be large; don't leak across iterations). The happy-path cleanup that pairs with
        `write_results`' TTL safety net."""
        out = []
        pipe = self.r.pipeline(transaction=False)
        order = []
        for (idx, n, fd, ns) in metas:
            if n == 0:
                continue
            xk, pik, mk, yk = result_keys(res_token, idx)
            pipe.get(xk); pipe.get(pik); pipe.get(mk); pipe.get(yk)
            order.append((idx, n, fd, ns, xk, pik, mk, yk))
        if not order:
            return out
        blobs = pipe.execute()
        # delete the result keys (raw bytes can be large; don't leak across iterations)
        dpipe = self.r.pipeline(transaction=False)
        for k, (idx, n, fd, ns, xk, pik, mk, yk) in enumerate(order):
            xb, pib, mb, yb = blobs[4 * k:4 * k + 4]
            X = np.frombuffer(xb, dtype=np.float32).reshape(n, fd)
            PI = np.frombuffer(pib, dtype=np.float32).reshape(n, ns)
            M = np.frombuffer(mb, dtype=np.float32).reshape(n, ns)
            Y = np.frombuffer(yb, dtype=np.float32)
            for i in range(n):
                out.append((X[i], PI[i], M[i], float(Y[i])))
            dpipe.delete(xk, pik, mk, yk)
        dpipe.execute()
        return out


def read_weights(conn, run, version):
    """Worker-side weight READ: fetch the raw-bytes payload for (run, version) over `conn` and return
    (manifest_str, blob_bytes). A missing payload is a loud RuntimeError (ADR-0002: never a silent
    stale-net serve). The bytes feed `unpack_net`; the (un)packing stays the WeightContainer's."""
    mk, bk = weight_keys(run, version)
    manifest = conn.get(mk)
    blob = conn.get(bk)
    if manifest is None or blob is None:
        raise RuntimeError(f"weight payload az:w:{run}:{version} missing from redis")
    return manifest.decode("utf-8"), blob


def write_results(conn, res_token, idx, X, PI, M, Y):
    """Worker-side result WRITE: pipeline the four contiguous float32 blocks (X/PI/M/Y) under the
    per-task result keys, each with the result TTL set in the same SET round-trip (the aborted-iteration
    self-clean safety net). No pickle — `tobytes()` of contiguous arrays."""
    xk, pik, mk, yk = result_keys(res_token, idx)
    ttl = _result_ttl()
    pipe = conn.pipeline(transaction=False)
    pipe.set(xk, X.tobytes(), ex=ttl)
    pipe.set(pik, PI.tobytes(), ex=ttl)
    pipe.set(mk, M.tobytes(), ex=ttl)
    pipe.set(yk, Y.tobytes(), ex=ttl)
    pipe.execute()

File: cpp/src/transport.cpp
// cpp/src/transport.cpp
// Purpose: the C++ redis wire client (hiredis) — the wire as the ONLY contract (ADR-0012 P7). See
//   transport.hpp. It mirrors chocofarm/az/transport.py + config.py byte-for-byte: the key
//   namespace, the manifest-driven float64 weight read (no hardcoded offsets — P1), the four
//   float32 result blocks (no second encoder — P7), and the CHOCO_REDIS_* connection contract
//   (default 127.0.0.1:6379 db0 — no hardcoded port; mirrors config.redis_params).
//
// Public Domain (The Unlicense).
#include "chocofarm/transport.hpp"

#include <cstdlib>
#include <cstring>
#include <stdexcept>

#include <hiredis/hiredis.h>
#include <nlohmann/json.hpp>

namespace chocofarm {

using nlohmann::json;

// ---- env contract (mirrors config.redis_params / config._result_ttl) ----
static std::string env_str(const char* name, const std::string& dflt) {
    const char* v = std::getenv(name);
    return v ? std::string(v) : dflt;
}
static int env_int(const char* name, int dflt) {
    const char* v = std::getenv(name);
    return v ? std::atoi(v) : dflt;
}
int result_ttl() { return env_int("CHOCO_RESULT_TTL", 3600); }  // mirrors transport._result_ttl

// ---- key namespace (byte-identical to transport.weight_keys / result_keys) ----
std::pair<std::string, std::string> weight_keys(const std::string& run, const std::string& phase,
                                                 int version) {
    std::string base = "az:w:" + run + ":" + phase + ":" + std::to_string(version);
    return {base + ":m", base + ":b"};
}
ResultKeys result_keys(const std::string& res_token, int idx) {
    std::string base = "az:res:" + res_token + ":" + std::to_string(idx);
    return ResultKeys{base + ":X", base + ":PI", base + ":M", base + ":Y"};
}

// ---- connection ----
RedisClient::RedisClient() {
    // CHOCO_REDIS_HOST / CHOCO_REDIS_PORT / CHOCO_REDIS_DB — the SAME env contract as config.py
    // (defaults 127.0.0.1 / 6379 / 0). No hardcoded port (ADR-0012 P7 / P1: config.py is the one
    // owner of "which redis"; this just reads the same env vars so it lands on the same instance).
    std::string host = env_str("CHOCO_REDIS_HOST", "127.0.0.1");
    int port = env_int("CHOCO_REDIS_PORT", 6379);
    int db = env_int("CHOCO_REDIS_DB", 0);

    ctx_ = redisConnect(host.c_str(), port);
    if (ctx_ == nullptr || ctx_->err) {
        std::string msg = ctx_ ? ctx_->errstr : "allocation failure";
        if (ctx_) { redisFree(ctx_); ctx_ = nullptr; }
        // ADR-0002 / P5: fail loud NOW if redis is unreachable, not mid-run (mirrors connect()'s ping).
        throw std::runtime_error("chocofarm: redis connect failed (" + host + ":" +
                                 std::to_string(port) + "): " + msg);
    }
    if (db != 0) {
        redisReply* r = static_cast<redisReply*>(redisCommand(ctx_, "SELECT %d", db));
        if (r == nullptr) throw std::runtime_error("chocofarm: redis SELECT failed (no reply)");
        bool ok = (r->type != REDIS_REPLY_ERROR);
        freeReplyObject(r);
        if (!ok) throw std::runtime_error("chocofarm: redis SELECT db failed");
    }
    // fail loud now if unreachable (ADR-0002), mirroring r.ping() in transport.connect()
    redisReply* pong = static_cast<redisReply*>(redisCommand(ctx_, "PING"));
    if (pong == nullptr) throw std::runtime_error("chocofarm: redis PING failed (no reply)");
    freeReplyObject(pong);
}

RedisClient::~RedisClient() {
    if (ctx_) redisFree(ctx_);
}

// Fetch a raw-bytes value; returns false (and leaves `out` empty) on a nil reply (a missing key).
static bool redis_get_bytes(redisContext* ctx, const std::string& key, std::string& out) {
    redisReply* r = static_cast<redisReply*>(redisCommand(ctx, "GET %b", key.data(), key.size()));
    if (r == nullptr) throw std::runtime_error("chocofarm: redis GET failed (no reply) for " + key);
    bool present = (r->type == REDIS_REPLY_STRING);
    if (present) out.assign(r->str, r->len);   // raw bytes (may contain NULs — use len, not strlen)
    bool nil = (r->type == REDIS_REPLY_NIL);
    freeReplyObject(r);
    if (!present && !nil) throw std::runtime_error("chocofarm: redis GET unexpected reply for " + key);
    return present;
}

WeightPayload RedisClient::read_weights(const std::string& run, const std::string& phase,
                                        int version) {
    auto [mk, bk] = weight_keys(run, phase, version);
    std::string manifest, blob;
    bool have_m = redis_get_bytes(ctx_, mk, manifest);
    bool have_b = redis_get_bytes(ctx_, bk, blob);
    if (!have_m || !have_b) {
        // mirrors read_weights' RuntimeError (ADR-0002 / P5: never a silent stale-net serve).
        throw std::runtime_error("weight payload az:w:" + run + ":" + phase + ":" +
                                 std::to_string(version) + " missing from redis");
    }

    json m = json::parse(manifest);
    WeightPayload p;
    p.in_dim = m.at("in_dim").get<int>();
    p.H = m.at("H").get<int>();
    // n_actions may be null in the manifest (a value-only net); coerce null -> -1.
    p.n_actions = m.at("n_actions").is_null() ? -1 : m.at("n_actions").get<int>();
    p.y_mean = m.at("y_mean").get<double>();
    p.y_std = m.at("y_std").get<double>();
    p.residual = m.value("residual", false);  // older manifests without residual -> OFF (P1)

    // Bind each weight by the MANIFEST's (off, len, shape, dtype) — NO hardcoded offsets (ADR-0012
    // P1: a hardcoded offset is the cross-language form of the three-writer feature-layout cancer).
    for (const auto& e : m.at("layout")) {
        WeightEntry we;
        we.name = e.at("name").get<std::string>();
        we.dtype = e.at("dtype").get<std::string>();
        we.off = e.at("off").get<int64_t>();
        we.len = e.at("len").get<int64_t>();
        for (const auto& s : e.at("shape")) we.shape.push_back(s.get<int64_t>());

        // The blob is contiguous float64 ('<f8') — match it exactly (weights are float64, results
        // float32). Reject anything else loudly rather than misread (Port/ACL: translate-and-
        // validate, do not coerce; ADR-0012 P2).
        if (we.dtype != "<f8") {
            throw std::runtime_error("chocofarm: unexpected weight dtype '" + we.dtype +
                                     "' (expected '<f8'); the weight blob is float64");
        }
        if (we.off < 0 || we.len < 0 || we.off + we.len > static_cast<int64_t>(blob.size())) {
            throw std::runtime_error("chocofarm: weight entry " + we.name +
                                     " (off,len) exceeds blob bounds");
        }
        int64_t count = we.len / static_cast<int64_t>(sizeof(double));
        std::vector<double> arr(static_cast<size_t>(count));
        std::memcpy(arr.data(), blob.data() + we.off, static_cast<size_t>(we.len));
        p.layout.push_back(std::move(we));
        p.weights.push_back(std::move(arr));
    }
    return p;
}

void RedisClient::write_results(const std::string& res_token, int idx,
                                const std::vector<float>& X, int n, int feat_dim,
                                const std::vector<float>& PI, const std::vector<float>& M,
                                const std::vector<float>& Y, int n_slots) {
    // Emit EXACTLY what np.frombuffer(...).reshape(...) decodes (no second encoder; ADR-0012 P7):
    //   X (n, feat_dim), PI (n, n_slots), M (n, n_slots), Y (n,) — contiguous little-endian float32,
    //   row-major. The float vectors are already laid out row-major, so tobytes == raw memcpy.
    ResultKeys k = result_keys(res_token, idx);
    int ttl = result_ttl();
    auto set_block = [&](const std::string& key, const float* data, size_t count) {
        const char* bytes = reinterpret_cast<const char*>(data);
        size_t nbytes = count * sizeof(float);
        // SET key <bytes> EX <ttl> — the TTL in the SAME round-trip (aborted-iteration self-clean).
        redisReply* r = static_cast<redisReply*>(redisCommand(
            ctx_, "SET %b %b EX %d", key.data(), key.size(), bytes, nbytes, ttl));
        if (r == nullptr) throw std::runtime_error("chocofarm: redis SET failed (no reply) for " + key);
        bool ok = (r->type == REDIS_REPLY_STATUS || r->type == REDIS_REPLY_STRING);
        freeReplyObject(r);
        if (!ok) throw std::runtime_error("chocofarm: redis SET failed for " + key);
    };
    // sanity: the blocks must be the row-major sizes the reshape expects (fail loud, ADR-0002).
    if (X.size() != static_cast<size_t>(n) * feat_dim ||
        PI.size() != static_cast<size_t>(n) * n_slots ||
        M.size() != static_cast<size_t>(n) * n_slots ||
        Y.size() != static_cast<size_t>(n)) {
        throw std::runtime_error("chocofarm: result block size mismatch (would corrupt reshape)");
    }
    set_block(k.X, X.data(), X.size());
    set_block(k.PI, PI.data(), PI.size());
    set_block(k.M, M.data(), M.size());
    set_block(k.Y, Y.data(), Y.size());
}

}  // namespace chocofarm
